Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellPaintingEventArgs.cs

```diff
@@ -0,0 +1,197 @@
+        public DataGridViewCellPaintingEventArgs(DataGridView dataGridView, Graphics graphics, Rectangle clipBounds, Rectangle cellBounds, int rowIndex, int columnIndex, DataGridViewElementStates cellState, object value, object formattedValue, string errorText, DataGridViewCellStyle cellStyle, DataGridViewAdvancedBorderStyle advancedBorderStyle, DataGridViewPaintParts paintParts)
+        {
+            Contract.Requires(dataGridView != null);
+            Contract.Requires(graphics != null);
+            Contract.Requires(cellStyle != null);
+            Contract.Requires((paintParts & ~DataGridViewPaintParts.All) == DataGridViewPaintParts.None);
+        }
+        
+        /// <summary>
+        /// Paints the specified parts of the cell for the area in the specified bounds.
+        /// </summary>
+        /// <param name="clipBounds">A <see cref="T:System.Drawing.Rectangle"/> that specifies the area of the <see cref="T:System.Windows.Forms.DataGridView"/> to be painted.</param><param name="paintParts">A bitwise combination of <see cref="T:System.Windows.Forms.DataGridViewPaintParts"/> values specifying the parts to paint.</param><exception cref="T:System.InvalidOperationException"><see cref="P:System.Windows.Forms.DataGridViewCellPaintingEventArgs.RowIndex"/> is less than -1 or greater than or equal to the number of rows in the <see cref="T:System.Windows.Forms.DataGridView"/> control.-or-<see cref="P:System.Windows.Forms.DataGridViewCellPaintingEventArgs.ColumnIndex"/> is less than -1 or greater than or equal to the number of columns in the <see cref="T:System.Windows.Forms.DataGridView"/> control.</exception>
+        public void Paint(Rectangle clipBounds, DataGridViewPaintParts paintParts)
+        {
+            Contract.Requires(!(this.RowIndex < -1));
```

Comment:
Why to use so weird precondition? Why not to use `Requires(RowIndex >= -1)` as always?